### PR TITLE
Overnight rescue: native ownership/lifetime hardening series + first Pass D/E module pass

### PR DESCRIPTION
**Pull request by sburdges-eng**
## Summary

Multi-session native hardening series on `hermes-overnight-rescue`: an exhaustive file-by-file ownership/lifetime rescue across `src/`, `src_penta-core/`, and the FFI boundaries (RAII for PyObject/JSON/queue-node lifetimes, idempotent worker-thread lifecycles, JUCE Viewport double-delete fixes), followed by the first module-scoped RAII pass under the newly approved architecture workbook (Pass D/E), and a ThreadSanitizer certification of the result. The full audit trail lives in `hermes_rescue_state.md` (59 numbered items); approved architecture decisions are recorded in `docs/handoffs/20260609_034047_caca5b.md`.

## Changes

- **Rescue series (items 1–53):** PyObject RAII across all Python bridges; DynamicObject allocation funneled through local helpers in serialization paths; restart-safe/idempotent worker lifecycles (RTLogger ×2, MLInterface, AudioEmotionRunner, InferenceThreadManager, AudioWorkerThread); LockFreeQueue sentinel/teardown guards; RTMemoryPool out-of-pool deallocate rejection; OSC pending-request cleanup + response-type matching; PreferenceBridge fallback JSON corruption fix; ScoreEntryPanel/MixerConsolePanel `setViewedComponent` double-delete fix; Intent IR FFI session-id coherence with validation moved outside the state mutex
- **Module pass (items 54–58), `src/ml/MultiModelProcessor`:** worker owned as concrete `std::unique_ptr<InferenceThread>` (no type-erased downcasts); `start()`/`stop()` serialized + idempotent with cooperative join (no force-kill that could orphan the processor mutex mid-inference); request snapshot taken before releasing the submit slot (closes torn-feature reads); `loaded_`/`enabled_`/`initialized_` made atomic for off-lock readers; new standalone 11-case lifecycle test incl. a concurrent start/stop torture that crashed the old code 100% deterministically (pure-virtual call)
- **TSan enablement (item 59):** `KMIDI_ENABLE_INTENT_IR` CMake option (default ON, FATAL_ERROR guard vs `BUILD_KELLY_FFI`) so sanitizer trees can link KellyCore test executables without the Rust staticlib, which exports `compiler_builtins` symbols and overflows the arm64 compact-unwind personality limit under TSan
- **Docs:** approved architecture workbook handoff (passes A/C/D/E/F); rescue journal

### Author

- [x] Code follows existing project conventions (see `docs/NAMING_CONVENTIONS.md`)
- [x] Tests added or updated for changed behaviour
- [x] No new lint or type-check warnings introduced
- [x] Documentation updated where applicable
- [x] No secrets, credentials, or absolute paths committed
- [x] RT-safety rules preserved (no heap allocations on audio thread)

### Reviewer

- [ ] Logic is correct and handles edge cases
- [ ] No security or input-validation regressions
- [ ] CI checks pass (Python tests, C++ build, lint, formatting)
- [ ] Changes are minimal and focused on the stated goal

## Testing

```
# ASan/UBSan gate (Debug): full standalone C++ suite
ctest --test-dir build-asan --output-on-failure   # 7/7 pass; new MultiModelProcessorTest torture 10/10 repeats

# TSan certification (KMIDI_ENABLE_TSAN=ON, KMIDI_ENABLE_INTENT_IR=OFF)
ctest --test-dir build-tsan --output-on-failure   # 7/7 pass, zero TSan reports (5/5 repeat runs of the ML pipeline test)

# Link/compile gates
cmake --build build-debug --target KellyCore KellyFFI   # clean
cd engine/intent_ir && cargo test                        # passing (per-series validation log in hermes_rescue_state.md)
```

`src/ml/MultiModelProcessor` changes were additionally reviewed by the C++ safety reviewer (verdict: approve; its one actionable finding — silent `startThread()` failure — is fixed in `ac757e9a`).

**Reviewer notes:** the two module files were wholesale reindented by the format-on-write hook — use `git diff -w` on `ac757e9a` for the semantic diff (~131 lines). Exported ABI, Intent IR semantics, and persistence formats are untouched per the agent-surface rules in the handoff; the recorded residual (scrubbing non-ABI symbols from `libintent_ir.a`) is deliberately left for a human-approved slice. No build tree compiles `src/plugin/` (plugins OFF), so plugin sources were not recompiled against the header change; in-tree consumers exercising the same public API compile clean.

## Related Issues

None — overnight hardening series; architecture context in `docs/handoffs/20260609_034047_caca5b.md`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **High Risk**
> Touches native lifetimes, CMake link ownership, FFI test targets, and many CI workflows; ABI/schema/persistence are claimed unchanged but the blast radius is large and plugin paths were not recompiled in-tree.
> 
> **Overview**
> This PR bundles the **overnight native rescue** with a large **repo/CI/docs alignment** pass, so the diff is wider than the `MultiModelProcessor` headline alone.
> 
> **Native / CMake:** Adds **`KMIDI_ENABLE_INTENT_IR`** (optional Rust staticlib link; required for `KellyFFI`) so TSan builds can omit Intent IR; tightens KellyCore assembly (e.g. drop `intent_ir_ffi` when IR is off), splits **`prrot_core`** and moves **penta OSC / VoiceLeading** ownership so headless `BUILD_KELLY_CORE=OFF` CI can link; enables **PIC**, Windows **`KellyFFI.def`** narrow exports, optional Qt on KellyFFI, and new **`runtime_contract_tests`** / **`MultiModelProcessorTest`** plus sanitizer instrumentation on standalone C++ tests.
> 
> **CI:** Cross-workflow fixes—**bash** on Windows, **MSVC dev cmd**, **`BUILD_PENTA_TESTS`** naming, **`httpx`/`httpx2`**, **`KMIDI_OFFLINE_BUILD`** offline gate, lcov ignore flags, recursive submodules, bug-taxonomy job deps.
> 
> **Docs / governance:** Replaces **`docs/ARCHITECTURE.md`** with the 2026 handoff set (module map, Intent IR, RT, FFI, persistence, agent surfaces), refreshes **README/BUILD/QUICK_START/BOOT** away from Tauri-as-center, adds drift audit/checklists/Hermes plans, and marks legacy **KmiDi_PROJECT** scripts as historical.
> 
> **Tooling:** New **`.cursor/rules/swarm_handoff.mdc`**, **`.gitattributes`** for byte-exact JUCE patches, expanded **`.gitignore`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 78a673f840c1d7d0270523a12ce9889a23294d74. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -28,6 +28,12 @@ option(BUILD_MIDI_CI_DAEMON "Build MIDI-CI microformat-to-SysEx daemon" OFF)
 option(KMIDI_ENABLE_ASAN "Enable AddressSanitizer + UndefinedBehaviorSanitizer" OFF)
 option(KMIDI_ENABLE_TSAN "Enable ThreadSanitizer (mutually exclusive with ASAN)" OFF)
 option(KMIDI_ENABLE_RTSAN "Enable LLVM RealtimeSanitizer (mutually exclusive with ASAN/TSAN; requires Clang 20+)" OFF)
+# Rust Intent IR staticlib. OFF exists for TSan certification builds: the
+# prebuilt archive leaks compiler_builtins symbols (__divdc3 etc.), so the
+# linker drags in Rust std and rust_eh_personality, overflowing the arm64
+# compact-unwind personality limit once TSan's runtime joins the link.
+# FFI builds require this ON (KellyFFI embeds the staticlib).
+option(KMIDI_ENABLE_INTENT_IR "Link the Rust Intent IR staticlib into KellyCore" ON)
```
[the remark this answers is not in the document]
Comment (sburdges-eng) [reply CMakeLists.txt:46]: Fixed in c5587a08: intent_ir_ffi.cpp is now excluded from KELLY_CORE_SOURCES when KMIDI_ENABLE_INTENT_IR=OFF, so the archive carries no dangling Rust FFI references. Verified locally: TSan (OFF) tree builds with zero intent_ir_ffi objects and the full test suite passes; default (ON) tree still compiles it.

File: src/biometric/BiometricInput.h [hunk 2]
```diff
@@ -144,19 +149,8 @@ class BiometricInput {
...
-  // Use void* to avoid incomplete type issues with unique_ptr in header
-  // Will be cast to proper types in implementation
-  void *healthKitBridge_;
-  void *fitbitBridge_;
-
-  // Streaming thread
-  std::thread streamingThread_;
-  std::atomic<bool> shouldStream_;
-  void streamingLoop();
+  std::unique_ptr<biometric::HealthKitBridge> healthKitBridge_;
```
[the remark this answers is not in the document]
Comment (sburdges-eng) [reply src/biometric/BiometricInput.h:152]: Confirmed and fixed in 0fff5abc — reproduced the incomplete-type error standalone via the compile DB (-fsyntax-only on the .cpp twin: 'cannot delete an incomplete type'). HealthKitBridge.h is now included unconditionally; the old #if HEALTHKIT_AVAILABLE guard could never be true since that macro is defined inside the header itself. The header self-stubs on non-Apple, and the macOS build (which selects the .mm twin) is unaffected.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -21,7 +21,7 @@ jobs:
     - name: Install dependencies
       run: |
         pip install numpy scipy torch librosa pyyaml
-        pip install pytest pytest-cov
+        pip install pytest pytest-cov httpx2
```
[the remark this answers is not in the document]
Comment (sburdges-eng) [reply .github/workflows/tests.yml:24]: Not taking the httpx2→httpx swap: the requirement is starlette-version-dependent, and the failing ubuntu job's log says verbatim 'RuntimeError: The starlette.testclient module requires the httpx2 package to be installed. You can install this with: $ pip install httpx2' (run 27253223147/job 80481930650). Newer starlette (>=1.2, what pip resolves on these jobs) moved TestClient to httpx2; the older-starlette jobs keep httpx (test.yml's comment documents that era). pyproject's dev extra now carries both httpx>=0.24 (runtime imports in the MCP servers) and httpx2>=2.0 (TestClient), so tests.yml installing httpx2 for its unit-test job matches what its resolved starlette actually imports. Swapping would reintroduce the 11 collection errors.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -34,7 +34,7 @@ jobs:
       run: |
         pip install -e .
         pip install numpy scipy torch librosa pyyaml
-        pip install pytest pytest-cov coverage
+        pip install pytest pytest-cov coverage httpx2
```
[the remark this answers is not in the document]
Comment (sburdges-eng) [reply .github/workflows/ci.yml:37]: Empirically settled by the latest run: with httpx2 installed (plus httpx via the dev extra where used), Python Unit Tests collected cleanly — 445 passed, 0 import errors (was 11 collection errors). httpx2 2.3.0 is the successor line (deps httpcore2/truststore) and the resolved starlette imports it first, falling back to httpx. The job's 2 remaining failures are unrelated TTL-latency test bugs in tests/unit/test_agent_latency.py with a fix already in flight. Leaving the installs as-is; if a job pins an older starlette that wants classic httpx, it installs that explicitly (test.yml does).

File: .github/workflows/pre-training-hardening.yml [hunk 5]
```diff
@@ -70,8 +71,8 @@ jobs:
         run: scripts/ci/install_juce_linux_deps.sh
       - name: Configure and build runtime contract tests
         run: |
-          cmake -S . -B build-ci-runtime -DOFFLINE_BUILD=OFF -DJUCE_EXTERNAL_ROOT= -DKELLY_ENABLE_TESTS=ON
+          cmake -S . -B build-ci-runtime -DKMIDI_OFFLINE_BUILD=OFF -DBUILD_TESTS=ON
```
[the remark this answers is not in the document]
Comment (sburdges-eng) [reply .github/workflows/pre-training-hardening.yml:85]: Adopted in spirit in f84e63e6: added dtolnay/rust-toolchain@stable to the job. One correction for the record: configure does not currently fail — engine/intent_ir/CMakeLists.txt auto-enables INTENT_IR_REBUILD_RUST when the prebuilt is absent and cargo exists, and the runner image ships cargo (the prior run's configure succeeded in 75s and reached the build step). The pin guards against runner-image drift rather than fixing a live break.

File: CMakeLists.txt [hunk 6]
```diff
@@ -289,6 +338,27 @@ list(FILTER KELLY_CORE_SOURCES EXCLUDE REGEX "/src/export/StemExporter\\.cpp$")
...
+# prrot:: ownership (2026-06-10): src/prrot/ builds as the standalone
+# prrot_core static lib (defined above) so penta_core_native links without
+# KellyCore. KellyCore keeps the symbols via its prrot_core link below.
+list(FILTER KELLY_CORE_SOURCES EXCLUDE REGEX "/src/prrot/.*\\.cpp$")
+
+# penta::harmony::VoiceLeading (2026-06-10): referenced by penta_core's own
+# HarmonyEngine.cpp — the remaining 2 LNK2001 externals in the same headless
+# job. The impl now compiles into penta_core (src_penta-core/CMakeLists.txt
+# appends it); KellyCore keeps the symbols via its penta_core link.
+list(FILTER KELLY_CORE_SOURCES EXCLUDE REGEX "/src/harmony/VoiceLeading\\.cpp$")
```
[the remark this answers is not in the document]
Comment (sburdges-eng) [reply CMakeLists.txt:360]: Confirmed and fixed in 46dfe32b — your diagnosis was exactly right: the push carried the root-side KELLY_CORE_SOURCES exclusions with a stale penta-side file, so the four TUs compiled into no library and every penta_core/KellyCore consumer collapsed. Declined the autofix's unconditional target_sources though: the OSC trio needs juce_osc (gated on TARGET juce::juce_osc with the module linked PUBLIC) so standalone no-JUCE builds keep the JUCE-free subset; VoiceLeading.cpp is pure-std and appended unconditionally. Verified in a fresh Release worktree: 0 build errors, ctest 15/15 including penta_tests and runtime_contract_tests.

File: CMakeLists.txt [hunk 7]
```diff
@@ -289,6 +338,27 @@ list(FILTER KELLY_CORE_SOURCES EXCLUDE REGEX "/src/export/StemExporter\\.cpp$")
...
+# windows-latest. KellyCore keeps the symbols via its penta_core link;
+# excluding here prevents duplicate definitions across the two archives.
+list(FILTER KELLY_CORE_SOURCES EXCLUDE REGEX "/src/osc/OSCClient\\.cpp$")
+list(FILTER KELLY_CORE_SOURCES EXCLUDE REGEX "/src/osc/OSCHub\\.cpp$")
+list(FILTER KELLY_CORE_SOURCES EXCLUDE REGEX "/src/osc/OSCServer\\.cpp$")
+
+# prrot:: ownership (2026-06-10): src/prrot/ builds as the standalone
+# prrot_core static lib (defined above) so penta_core_native links without
+# KellyCore. KellyCore keeps the symbols via its prrot_core link below.
+list(FILTER KELLY_CORE_SOURCES EXCLUDE REGEX "/src/prrot/.*\\.cpp$")
```
[the remark this answers is not in the document]
Comment (sburdges-eng) [reply CMakeLists.txt:354]: Confirmed and fixed in 78a673f8: added prrot_core to penta_core_native's PRIVATE links (matching your autofix, with a comment on why the edge exists). Verified locally on the headless shape — BUILD_KELLY_CORE=OFF configure builds and links the module cleanly, which previously was exactly the unresolved-prrot failure.
